### PR TITLE
Make the SLURM Preemption/Timeout Signal Configurable

### LINKED ISSUE #14610: Make the Slurm Signal Configurable
## 🚀 Feature

Lightning currently assumes that the slurm preemption signal is set to USR1, it would be nice to add an option to SLURMEnvironment to change this.

### Motivation

I use submitit to handle slurm submissions and they recently moved away from USR1 because of an NCCL conflict.

This breaks the requeuing in pytorch lightning. 

See also https://github.com/facebookincubator/submitit/issues/1709

### Pitch

`SLURMEnvironment` can be trivially extended to account for this with something like 

```
def __init__(self, auto_requeue: bool = True, signal: str = "USR1") -> None:
        super().__init__()
        self.auto_requeue = auto_requeue
        self.signal = signal
```

then `SignalConnector` ~ln 60 can be changed to something like:

```
sig = getattr(signal, "SIG" + environment.signal, None)
self._register_signal(sig, HandlersCompose(sigusr1_handlers))
```

cc @borda @awaelchli

### PR DESCRIPTION
**Pull request by Queuecumber**
## What does this PR do?

This PR adds a flag to the `SLURMEnvironment` class which allows users to change which signal will be watched for preemtion/timeout. 

Although SIGUSR1 is fairly standard, some libraries (e.g., submitit) which submit to SLURM use a different signal or users may have another compatibility issue with SIGUSR1 and therefore require the use of another signal. 

Fixes #14610 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Queuecumber** [comment]
@awaelchli This is ready for review, I have two failing tests but I'm not sure whats causing it

**Comment by Queuecumber** [comment]
One thing to discuss: 

The way I implemented this if the user specifies an invalid signal, that emits a warning message and the SignalConnector falls back to using SIGUSR1 

This could also be a fatal error of some sort (either exception and crash or error log and no signal listening)

File: docs/source-pytorch/clouds/cluster_advanced.rst [hunk 1]
```diff
@@ -88,6 +88,18 @@ To train a model using multiple nodes, do the following:
 
         #SBATCH --signal=SIGUSR1@90
 
+    You can change this signal if your environment requires the use of a different one, for example
+
+    .. code-block:: bash
+
+        #SBATCH --signal=SIGHUP@90
+
+    Then, when you make your trainer, pass the ` signal` option to the :class:`~pytorch_lightning.plugins.environments.slurm_environment.SLURMEnvironment` plugin::
```
Comment (awaelchli) [inline docs/source-pytorch/clouds/cluster_advanced.rst:97]:
```suggestion
    Then, when you make your trainer, pass the `signal` option to the :class:`~pytorch_lightning.plugins.environments.slurm_environment.SLURMEnvironment` plugin:
```

File: src/pytorch_lightning/plugins/environments/slurm_environment.py [hunk 2]
```diff
@@ -30,9 +30,10 @@ class SLURMEnvironment(ClusterEnvironment):
             rescheduled gets determined by the owner of this plugin.
     """
 
-    def __init__(self, auto_requeue: bool = True) -> None:
+    def __init__(self, auto_requeue: bool = True, signal: str = "USR1") -> None:
```
Comment (awaelchli) [inline src/pytorch_lightning/plugins/environments/slurm_environment.py:33]: Some considerations:
- We could support int here, i.e., signal: int = signal.SIGUSR1
- We could support both, i.e., signal: int | str = signal.SIGUSR1
- Is excluding the SIG prefix intuitive? Intuitively, I would have passed in `signal.SIGUSR1` or `"SIGUSR1"`. 

No strong opinions here, maybe I'm overthinking this.

File: src/pytorch_lightning/plugins/environments/slurm_environment.py [hunk 3]
```diff
@@ -30,9 +30,10 @@ class SLURMEnvironment(ClusterEnvironment):
             rescheduled gets determined by the owner of this plugin.
```
Comment (awaelchli) [inline src/pytorch_lightning/plugins/environments/slurm_environment.py:30]: Would you mind also adding the docstring for the new signal argument? Thanks :)

File: src/pytorch_lightning/trainer/connectors/signal_connector.py [hunk 4]
```diff
@@ -51,19 +51,28 @@ def register_signal_handlers(self) -> None:
...
+        # Windows seems to have signal incompatibilities
         if not self._is_on_windows():
-            if sigusr1_handlers and not self._has_already_handler(signal.SIGUSR1):
-                self._register_signal(signal.SIGUSR1, HandlersCompose(sigusr1_handlers))
+            if isinstance(environment, SLURMEnvironment):
+                sigusr = getattr(signal, "SIG" + environment.signal, None)
+
+                if sigusr is None:
+                    log.warning(f"Unrecognized signal name SIG{environment.signal}, assuming SIGUSR1")
+                    sigusr = signal.SIGUSR1
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/signal_connector.py:64]: I suggest to move this validation to the SLURMEnvironment directly. What do you think?
Then, we can assume that SLURMEnvironment.signal will always return a good one.

File: src/pytorch_lightning/trainer/connectors/signal_connector.py [hunk 5]
```diff
@@ -51,19 +51,28 @@ def register_signal_handlers(self) -> None:
...
+            sigusr_handlers.append(self.slurm_sigusr_handler_fn)
             sigterm_handlers.append(self.sigterm_handler_fn)
 
-        # signal.SIGUSR1 doesn't seem available on windows
+        # Windows seems to have signal incompatibilities
         if not self._is_on_windows():
-            if sigusr1_handlers and not self._has_already_handler(signal.SIGUSR1):
-                self._register_signal(signal.SIGUSR1, HandlersCompose(sigusr1_handlers))
+            if isinstance(environment, SLURMEnvironment):
+                sigusr = getattr(signal, "SIG" + environment.signal, None)
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/signal_connector.py:60]: See my considerations above. I think this logic could be simplified, by direct access via environment.signal, and the default (sigusr = signal.SIGUSR1) should also be set in the SLURMEnvironment IMO.

**Review by awaelchli — COMMENTED** [review]
Overall approach and implementation looks good to me. Minor comments. Thanks for this contribution, and the testing on the cluster!

**Comment by awaelchli** [comment]
> The way I implemented this if the user specifies an invalid signal, we emit a warning message and the SignalConnector falls back to using SIGUSR1

I think we should immediately check this in the init of SLURMEnvironment. Did you choose a warning for a reason? I.e., long waiting job in the queue and erroring out can be annoying to the user? If there is a concern, I think that's fine. Otherwise I would error out, as warnings could be easily overlooked or lost in logs.

**Comment by awaelchli** [comment]
> I have two failing tests but I'm not sure whats causing it

Will look at it once we have resolved the discussions here. Probably a difference in the signal library between Python versions.

File: src/pytorch_lightning/plugins/environments/slurm_environment.py (hunk 3 above)
Comment (Queuecumber) [reply src/pytorch_lightning/plugins/environments/slurm_environment.py:30]: Yeah completely forgot that

File: src/pytorch_lightning/plugins/environments/slurm_environment.py (hunk 2 above)
Comment (Queuecumber) [reply src/pytorch_lightning/plugins/environments/slurm_environment.py:33]: I think going straight for `signal: int = signal.SIGUSR1` makes sense actually, and this simplifies the code quite a bit so less room for error, I'll make that change. I don't see any particular advantage for the string parameter.

File: tests/tests_pytorch/trainer/connectors/test_signal_connector.py [hunk 6]
```diff
@@ -90,14 +90,37 @@ def test_auto_requeue_flag(auto_requeue):
...
     else:
         assert signal.getsignal(signal.SIGTERM) is signal.SIG_DFL
         assert signal.getsignal(signal.SIGUSR1) is signal.SIG_DFL
 
     connector.teardown()
 
 
+@RunIf(skip_windows=True)
+@pytest.mark.parametrize("auto_requeue", (True, False))
+@pytest.mark.parametrize("requeue_signal", [(signal.SIGUSR1, signal.SIGUSR2, signal.SIGHUP)])
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/connectors/test_signal_connector.py:103]: I wouldn't duplicate the test above just for checking this parametrization. The difference is only a single line, why not merge the tests?

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/trainer/connectors/test_signal_connector.py (hunk 6 above)
Comment (Queuecumber) [reply tests/tests_pytorch/trainer/connectors/test_signal_connector.py:81]: Initially it was to make sure the default argument was working as expected but that's not as important with the updated interface from the above discussions

File: docs/source-pytorch/clouds/cluster_advanced.rst [hunk 7]
```diff
@@ -88,6 +88,18 @@ To train a model using multiple nodes, do the following:
 
         #SBATCH --signal=SIGUSR1@90
 
+    You can change this signal if your environment requires the use of a different one, for example
```
Comment (awaelchli) [inline docs/source-pytorch/clouds/cluster_advanced.rst:91]: This step by step guide here is to show how to get started quickly with default settings. So I think it's better not to talk about customization here directly. 

I would move this new paragraph to the detail description in the section "Enable auto wall-time resubmitions" below. What do you think?

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/trainer/connectors/test_signal_connector.py [hunk 8]
```diff
@@ -78,8 +78,9 @@ def training_step(self, batch, batch_idx):
 
 @RunIf(skip_windows=True)
 @pytest.mark.parametrize("auto_requeue", (True, False))
-def test_auto_requeue_flag(auto_requeue):
-    trainer = Trainer(plugins=[SLURMEnvironment(auto_requeue=auto_requeue)])
+@pytest.mark.parametrize("requeue_signal", [(signal.SIGUSR1, signal.SIGUSR2, signal.SIGHUP)])
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/connectors/test_signal_connector.py:81]:
```suggestion
@pytest.mark.parametrize("requeue_signal", [signal.SIGUSR1, signal.SIGUSR2, signal.SIGHUP])
```

This will fix the "expected int, not tuple" test error

Comment (Queuecumber) [reply tests/tests_pytorch/trainer/connectors/test_signal_connector.py]: Oops missed that

**Review by justusschock — APPROVED** [review]

**Comment by Queuecumber** [comment]
This is rebased on latest master and ready to go unless there is more feedback

File: src/pytorch_lightning/trainer/connectors/signal_connector.py [hunk 9]
```diff
@@ -51,19 +51,24 @@ def register_signal_handlers(self) -> None:
...
 
-        # signal.SIGUSR1 doesn't seem available on windows
+        # Windows seems to have signal incompatibilities
         if not self._is_on_windows():
-            if sigusr1_handlers and not self._has_already_handler(signal.SIGUSR1):
-                self._register_signal(signal.SIGUSR1, HandlersCompose(sigusr1_handlers))
+            if isinstance(environment, SLURMEnvironment):
+                sigusr = environment.requeue_signal
+            else:
+                sigusr = signal.SIGUSR1
```
Comment (Borda) [inline src/pytorch_lightning/trainer/connectors/signal_connector.py:62]:
```suggestion
            sigusr = environment.requeue_signal if isinstance(environment, SLURMEnvironment) elsesignal.SIGUSR1
```

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@awaelchli seems to be ready... :)

**Comment by awaelchli** [comment]
Earlier I promised I would take care of the failing Windows tests: https://github.com/Lightning-AI/lightning/actions/runs/3039218876/jobs/4893893890

Which is what I'm doing now. Otherwise we can't merge with failing tests.

**Comment by Queuecumber** [comment]
Ok I understand now, because python on windows doesn't even define those constants for the signals

**Review by rohitgr7 — APPROVED** [review]

**Comment by awaelchli** [comment]
@Queuecumber Looks like we got the Windows cases figured out. 
Great work, thanks for contributing this improvement!

**Comment by Queuecumber** [comment]
Thanks for merging, do you guys know what the ETA is for it showing up in an installable package (pre-release or full release)

I'm eager to use this in semi-production

**Comment by awaelchli** [comment]
This feature will go in to 1.8, we are targeting end of September / beginning October. 

To use this immediately, you can install the build from the master branch: 
pip install https://github.com/Lightning-AI/lightning/archive/refs/heads/master.zip -U
